### PR TITLE
[ProofOfConcept, WIP] Features/parameter highlighting enhancements

### PR DESCRIPTION
**Pull request by En3Tho**
Just a proof of concept. These changes can be used to enable function parameter highlighting (example is from JetBrains Rider):
image
image

Note that there are some caveats, distinction of top level lambdas and nested scope lambdas (which is used for highlighting) is not obvious in some cases

Need review
Need some help with adding tests for this if accepted

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=10327) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=10327) before we can accept your contribution.<br/><br/>:x: En3Tho [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=10327)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=10327) it.</sub>

File: src/fsharp/symbols/Symbols.fsi [hunk 1]
```diff
@@ -699,6 +699,15 @@ and [<Class>] public FSharpMemberOrFunctionOrValue =
     internal new : SymbolEnv * ValRef -> FSharpMemberOrFunctionOrValue
     internal new : SymbolEnv * Infos.MethInfo -> FSharpMemberOrFunctionOrValue
 
+    /// Indicates if this is a top level function parameter (like x in let f x = ... or x in fun x -> ...)
+    member IsParameter : bool
```
Comment (cartermp) [inline src/fsharp/symbols/Symbols.fsi:703]: Suggested rename: `IsFunctionParameter`, same with the others

**Review by cartermp — COMMENTED** [review]
Very cool! I love to see this kind of distinguishing work. It may also come in handy for

Something worth considering is that consumers will have to rationalize this vs. the `CurriedParameterGroups` that constructs a nested list of `FSharpParameter`. I personally found it kind of...useless for https://github.com/dotnet/fsharp/pull/10295/ and had to resort to using a different way to get argument names and associate them with their definition the way that I wanted, though. Maybe there's future work that could be done to make `FSharpParameter` more useful?

As for tests, here's where they would go: https://github.com/dotnet/fsharp/blob/main/tests/service/Symbols.fs

I would think that just having some source code as a string and then asserting some things about symbols you try to find from the symbol results would be sufficient. The Symbol API tests are pretty barren, as you can see

**Comment by En3Tho** [comment]
@cartermp While I was enabling new highlighting options in Rider 2020.2 I found that FSharpParameter is a very strange creature indeed. I tried to highlight SRTP dependencies (distinguish methods/functions and properties for readability) but haven't found a proper way to distinguish them via this API. Also, enabling this kind of highlighting (with locals and parameters separated) for ctor parameters using FSharpParameter proved to be hard too because it highlighted those both at type declaration and ctor call site

**Comment by cartermp** [comment]
Conflicts + some test failures for baselines. Additionally I do think that the names need a change. But other than that I think this is fine.

In the long-term we might try to make FSharpParameter more useful and find a way to re-use work here. But I'm prepared to take this change with the more specific naming.

**Comment by cartermp** [comment]
Given some of the feedback I had in this area in my other PR, I think @dsyme should weigh in on the shape of the API here

**Comment by En3Tho** [comment]
@cartermp Well, yep. I did some renaming as you suggested and I quite like it, but it's verbose (I personally like this because it's easily readable, but... ->) and doesn't really fit in the compiler's code which is full of abbreviations.

**Comment by En3Tho** [comment]
@dsyme Can you please look at this and tell us if you like it or not? Any suggestions are welcome!

**Comment by En3Tho** [comment]
@cartermp @dsyme If I understood Don's opinion correctly, the thing I'm trying to implement here goes with accordance to Don's thoughts about distinction of top level vs limited scope values (also, locals)

> Values - top level v. non-top-level (open-scope v. limited scope)
> 
> Don's opinion: I can see the value for a subtle distinction - among other things it likely helps to orient the programmer when they are in a large module v. a large class/function body
> 
> Explainer: Right now there is a fairly subtle distinction between Module/member and class/expression bindings. I don't have strong feelings about this as the distinctions are very subtle. But it's worth noticing that it's not distinguishing static. v. non-static, in the following I'm seeing xxxxxxxxxxxxxxxx and yyyyyyyyyyyyyyyy get different colors.

For examle take this code from compiler:
I personally think it looks better and helps me distinguish things with very little cognitive load (I don't have to think where this or that thing comes from - color is usually enough):
Without:
image
With:
image

But again, there not really ideal situations when function parameter is actually a deconstructed single DU value:
Without:
image
With
image

In the end I think it comes down to personal preference, maybe even some kind biological difference in perception. Some of my colleagues used simple black color all over the place and to them my color scheme was too distracting. And to me their color scheme was simply unusuable, looking like a simple notepad.

I would like to somehow proceed with this change (need review from Don) or close it?

**Comment by En3Tho** [comment]
Lots of white space changes after merge, seems like IDE automatically deleted them because it assumed those lines were changed by me. I hope this is ok.

**Comment by cartermp** [comment]
Just to re-affirm my support for this API - I do feel like we just need additional input though.

**Comment by En3Tho** [comment]
Pinging @dsyme again...

File: src/fsharp/TypedTreePickle.fs [hunk 2]
```diff
@@ -2312,6 +2312,7 @@ and u_ValData st =
       val_type         = x2
       val_stamp        = newStamp()
       val_flags        = ValFlags x4
+      val_flags_2      = ValFlags2 NotAFunctionParameter
```
Comment (cartermp) [inline src/fsharp/TypedTreePickle.fs:2315]: @TIHan @dsyme does this add to the metadata blob?

File: src/fsharp/symbols/Symbols.fsi [hunk 3]
```diff
@@ -721,6 +721,15 @@ type FSharpMemberOrFunctionOrValue =
     internal new: SymbolEnv * ValRef -> FSharpMemberOrFunctionOrValue
     internal new: SymbolEnv * Infos.MethInfo -> FSharpMemberOrFunctionOrValue
 
+    /// Indicates if this is a top level function parameter (like x in let f x = ... or x in fun x -> ...)
```
Comment (cartermp) [inline src/fsharp/symbols/Symbols.fsi:724]: this comment and the one below it are identical. Is that intended?

**Review by cartermp — CHANGES_REQUESTED** [review]
Question about the metadata format that @TIHan or @dsyme is going to have to answer. If this does add to the metadata format, we'll need to consider this pretty deeply. It's technically a forwards-incompatible change if that is the case, and one that is nearly impossible to understand from an error messaage lone.

File: src/fsharp/symbols/Symbols.fsi (hunk 3 above)
Comment (En3Tho) [reply src/fsharp/symbols/Symbols.fsi:724]: Not really, it's more about not knowing what to write as there is lot's of duplication of that property

Comment (En3Tho) [reply src/fsharp/symbols/Symbols.fsi:724]: Oh, no. This in particular is my mistake.

**Comment by En3Tho** [comment]
seems like this pr is cursed, haha

**Comment by En3Tho** [comment]
@dsyme Pinging Don Syme again :) We really need your take on this one!

**Comment by En3Tho** [comment]
@dsyme Ping again :)

**Comment by abelbraaksma** [comment]
Fwiw, this is a great addition! You might wanna check as well what happens when identifiers are shadowed. The odd coloring in let bang syntax will actually be quite helpful, as it currently requires you to spot the excl mark, which isn't always very clear. Extra coloring for the lh side variable is then a better visual clue.

Hope @dsyme can soon take a look! :)

**Comment by En3Tho** [comment]
2 weeks have passed and here goes another ping! @dsyme . Please help :)

**Pull request closed without merging**